internal/stage_hp2.go

```diff
@@ -0,0 +1,98 @@
+
+package internal
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/codecrafters-io/shell-tester/internal/logged_shell_asserter"
+	"github.com/codecrafters-io/shell-tester/internal/shell_executable"
+	"github.com/codecrafters-io/shell-tester/internal/test_cases"
+	"github.com/codecrafters-io/shell-tester/internal/utils"
+	"github.com/codecrafters-io/tester-utils/random"
+	"github.com/codecrafters-io/tester-utils/test_case_harness"
+)
+
+func testHP2(stageHarness *test_case_harness.TestCaseHarness) error {
+	logger := stageHarness.Logger
+	shell := shell_executable.NewShellExecutable(stageHarness)
+	asserter := logged_shell_asserter.NewLoggedShellAsserter(shell)
+
+	// Create temporary history file paths
+	historyFile := filepath.Join(os.TempDir(), random.RandomWord()+"_shell_history_test")
+	defer os.Remove(historyFile) // Clean up the history file when done
+
+	// Set HISTFILE to /dev/null before starting the shell
+	shell.Setenv("HISTFILE", "/dev/null")
+
+	if err := asserter.StartShellAndAssertPrompt(true); err != nil {
+		return err
+	}
+
+	// Generate some random commands
+	nCommands := 3
+	commandTestCases := make([]test_cases.CommandResponseTestCase, nCommands)
+	for i := 0; i < nCommands; i++ {
+		randomWords := strings.Join(random.RandomWords(2), " ")
+		commandTestCases[i] = test_cases.CommandResponseTestCase{
+			Command:        "echo " + randomWords,
+			ExpectedOutput: randomWords,
+			SuccessMessage: "✓ Received expected response",
+		}
+	}
+
+	// Execute initial commands
+	for _, command := range commandTestCases {
+		if err := command.Run(asserter, shell, logger); err != nil {
+			return err
+		}
+	}
+
+	// Write history to first file
+	historyWriteCommand := "history -w " + historyFile
+	logger.Infof("Writing history to file using command: %s", historyWriteCommand)
+
+	historyTestCase := test_cases.CommandReflectionTestCase{
+		Command:        historyWriteCommand,
+		SuccessMessage: "✓ History -w command executed",
+	}
+	if err := historyTestCase.Run(asserter, shell, logger, false); err != nil {
+		return err
+	}
+
+	// Verify first history file contents
+	logger.Infof("Verifying history file contents")
+	historyContent, err := os.ReadFile(historyFile)
+	if err != nil {
+		logger.Errorf("Failed to read history file: %v", err)
+		return err
+	}
+
+	utils.LogReadableFileContents(logger, string(historyContent), "History file content:", filepath.Base(historyFile))
+
+	// Check if all commands are present in the history file in order, allowing for number prefixes
+	historyStr := strings.TrimSpace(string(historyContent))
+	expectedCommands := make([]string, nCommands+1)
+	for i, cmd := range commandTestCases {
+		expectedCommands[i] = cmd.Command
+	}
+	historyLines := strings.Split(historyStr, "\n")
+	if len(historyLines) != len(expectedCommands) {
+		return fmt.Errorf("history file has %d lines, expected %d", len(historyLines), len(expectedCommands))
```

Comment:
This won't be a super informative error message since the user might not know what we expect. Let's assert line by line and then complain (if the first few lines are correct let's say that first and then say "oh lines are missing at the end)

Reply:
The ideal error message tells you exactly what to do next